lib/llvmopencl/FixUnreachable.cc

```diff
@@ -0,0 +1,146 @@
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+#include "CompilerWarnings.h"
+IGNORE_COMPILER_WARNING("-Wmaybe-uninitialized")
+#include <llvm/ADT/Twine.h>
+POP_COMPILER_DIAGS
+IGNORE_COMPILER_WARNING("-Wunused-parameter")
+#include <llvm/IR/Constants.h>
+#include <llvm/IR/IRBuilder.h>
+#include <llvm/IR/Instructions.h>
+
+#include "FixUnreachable.h"
+#include "LLVMUtils.h"
+#include "WorkitemHandlerChooser.h"
+POP_COMPILER_DIAGS
+
+#include <iostream>
+#include <map>
+#include <set>
+
+#define PASS_NAME "fix-unreachable"
+#define PASS_CLASS pocl::FixUnreachable
+#define PASS_DESC "Fix unreachable instruction uses to defined behavior"
+
+// #define DEBUG_FIX_UNREACHABLE
+
+namespace pocl {
+
+using namespace llvm;
+
+static bool fixUnreachable(Function &F) {
+
+  Module *M = F.getParent();
+
+#ifdef DEBUG_FIX_UNREACHABLE
+  std::cerr << "@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@ BEFORE fixUnreachable: \n";
+  M->dump();
+  std::cerr << "@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@ \n";
+#endif
+
+  SmallVector<Instruction *, 8> PendingUnreachableInst;
+  SmallVector<BasicBlock *, 8> PendingDeletableBBs;
+  for (Function::iterator I = F.begin(), E = F.end(); I != E; ++I) {
+    for (BasicBlock::iterator BI = I->begin(), BE = I->end(); BI != BE;) {
+      if (auto UI = dyn_cast<UnreachableInst>(BI++)) {
+#ifdef DEBUG_FIX_UNREACHABLE
+        std::cerr << "UNREACHABLE found: replacing Inst in "
+                  << F.getName().str() << "\n";
+#endif
+        BasicBlock *BB = UI->getParent();
+        // this can happen when inlining functions which have unreachable Inst
+        // we end up with a BB with 0 predecessors and a single unreachable
+        if (BB->hasNPredecessors(0)) {
+          PendingDeletableBBs.push_back(BB);
+        } else {
+          PendingUnreachableInst.push_back(UI);
+        }
+      }
+    }
+  }
+
+  if (PendingUnreachableInst.empty())
+    return false;
+
+  for (auto BB : PendingDeletableBBs) {
+    BB->eraseFromParent();
+  }
+
+  // Find basic block with return instruction
+  BasicBlock *RetBB = nullptr;
+  for (Function::iterator I = F.begin(), E = F.end(); I != E; ++I) {
+    BasicBlock *BB = &*I;
+    if (BB->getTerminator() == nullptr)
```

Comment:
This should only happen if the IR is ill-formed, I think. All BBs should end with a terminator instruction.

Reply:
converted to assert()